wsd/ClientRequestDispatcher.cpp

```diff
@@ -990,6 +1003,155 @@ void ClientRequestDispatcher::handleWopiDiscoveryRequest(
     LOG_INF("Sent discovery.xml successfully.");
 }
 
+void ClientRequestDispatcher::handleWopiAccessCheckRequest(const Poco::Net::HTTPRequest& request,
+                                                           Poco::MemoryInputStream& message,
+                                                           const std::shared_ptr<StreamSocket>& socket)
+{
+    assert(socket && "Must have a valid socket");
+
+    LOG_DBG("Wopi Access Check request: " << request.getURI());
+
+    // Poco::JSON
+    Poco::JSON::Parser jsonParser;
+    Poco::Dynamic::Var parsingResult;
+    Poco::JSON::Object::Ptr object;
+    Poco::URI uri;
+
+    std::string text(std::istreambuf_iterator<char>(message), {});
+    try
+    {
+        parsingResult = jsonParser.parse(text);
+        object = parsingResult.extract<Poco::JSON::Object::Ptr>();
+
+        auto callbackUrlStr = object->getValue<std::string>("callbackUrl");
+
+        uri = Poco::URI(callbackUrlStr);
+    }
+    catch (const std::exception& exception)
+    {
+        LOG_ERR_S("Wopi Access Check request error, json object expected got ["
+                  << text << "] on request to URL: " << request.getURI() << exception.what());
+
+        HttpHelper::sendErrorAndShutdown(http::StatusCode::BadRequest, socket);
+        return;
+    }
+
+    enum class CheckStatus
+    {
+        Ok = 0,
+        NotHttpSucess,
+        HostNotFound,
+        HostUnReachable,
+        UnspecifiedError,
+        ConnectionAborted,
+        ConnectionRefused,
+        InvalidCertificate,
+        CertificateValidation,
+        NotHttps,
+        NoScheme,
+        Timeout,
+    };
+    std::map<CheckStatus, std::string> checkStatusNames = {
+        { CheckStatus::Ok, "Ok" },
+        { CheckStatus::NotHttpSucess, "NOT_HTTP_SUCESS" },
+        { CheckStatus::HostNotFound, "HOST_NOT_FOUND" },
+        { CheckStatus::HostUnReachable, "HOST_UNREACHABLE" },
+        { CheckStatus::UnspecifiedError, "UNSPECIFIED_ERROR" },
+        { CheckStatus::ConnectionAborted, "CONNECTION_ABORTED" },
+        { CheckStatus::ConnectionRefused, "CONNECTION_REFUSED" },
+        { CheckStatus::InvalidCertificate, "INVALID_CERTIFICATE" },
+        { CheckStatus::NotHttps, "NOT_HTTPS" },
+        { CheckStatus::NoScheme, "NO_SCHEME" },
+        { CheckStatus::Timeout, "TIMEOUT" }
+    };
+
+    CheckStatus result = CheckStatus::Ok;
+
+    if (uri.getScheme().empty())
+    {
+        result = CheckStatus::NoScheme;
+    }
+    else if (uri.getScheme() != "https")
+    {
+        result = CheckStatus::NotHttps;
+    }
+    else
+    {
+        // request the url
+        try
+        {
+            const auto hostAddress(Poco::Net::DNS::resolve(uri.getHost()));
```

Comment:
This is synchronous and blocks the thread - poke Caolan to see how to use his caching async DNS thread.